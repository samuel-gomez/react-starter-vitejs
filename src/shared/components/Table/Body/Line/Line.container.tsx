import type { Tcol } from '../Body';
import Line, { TLine } from './Line';

export type TLineContainer = Omit<TLine, 'columns'> & {
  LineCmpt?: typeof Line;
  cols: [string, Tcol][];
  lineNumber: number;
};

const LineContainer = ({ LineCmpt = Line, cols, ...restLineProps }: TLineContainer) => {
  const columns = cols.map(([keyCol, value]) => ({
    keyCol,
    ...value,
  }));
  return <LineCmpt {...restLineProps} columns={columns} />;
};

export default LineContainer;
