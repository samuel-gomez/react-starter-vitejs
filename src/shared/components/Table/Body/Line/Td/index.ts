export { default } from './Td';
export type { TTdContainer } from './Td';
