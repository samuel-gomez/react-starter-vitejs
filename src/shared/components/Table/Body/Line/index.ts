export { default } from './Line.container';
