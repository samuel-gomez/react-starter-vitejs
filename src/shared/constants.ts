import packageJson from '../../package.json';

/** ***********************************************************************
 * FORM CONSTANTS
 ************************************************************************ */
export const MSG_REQUIRED = 'Le champ est obligatoire';

export const MSG_LENGTH_3 = 'Le champ doit faire au moins 3 caractères';
export const MSG_FORMAT = 'Format invalide';
export const MSG_ERROR_FORMAT = 'Format de fichier incorrect';

export const DEFAULT_OPTION_LABEL = '- Sélectionner -';
export const DEFAULT_OPTION = {
  value: '',
  label: DEFAULT_OPTION_LABEL,
};

// patterns
export const patternEmail = /^[\w-+]+@([\w-+]+\.)+[\w-]{2,}$/;
export const patternPhoneNumber = /^[+0][\d ]*$/;
export const patternZipCode = /^\d{5}$/;
export const patternSiret = /^\d{14}$/;
export const patternSiren = /^\d{9}$/;
export const patternSiretSiren = /^\d{9,14}$/;

export const phoneNumberRule = {
  required: true,
  minLength: 4,
  maxLength: 20,
  pattern: patternPhoneNumber,
};

export const emailRule = {
  required: true,
  minLength: 6,
  maxLength: 60,
  pattern: patternEmail,
};

/** ***********************************************************************
 * API STATUS CONSTANTS
 ************************************************************************ */
export const STATUS_HTTP = {
  SUCCESS: 200,
  CREATED: 201,
  BAD_REQUEST: 400,
  UNAUTHORIZE: 401,
  FORBIDDEN: 403,
  NOTFOUND: 404,
  SERVER_ERROR: 500,
};

export const STATUS_API = {
  SUCCESS: STATUS_HTTP.SUCCESS,
  CREATED: STATUS_HTTP.CREATED,
  WARNING: 40,
  ERROR: 50,
};

/** ***********************************************************************
 * API URL
 ************************************************************************ */
export const API_URL = {
  BASE: 'base',
  GITHUB: 'github',
};

export const STATUS_HTTP_MESSAGES = {
  [STATUS_HTTP.SUCCESS]: 'Succès: La requête a réussi',
  [STATUS_HTTP.CREATED]: 'Succès: Création réussie',
  [STATUS_HTTP.BAD_REQUEST]: 'Erreur: La syntaxe de la requête est erronée',
  [STATUS_HTTP.UNAUTHORIZE]: 'Erreur: Une authentification est nécessaire pour accéder à la ressource',
  [STATUS_HTTP.FORBIDDEN]: 'Erreur: Accès non autorisé',
  [STATUS_HTTP.NOTFOUND]: 'Erreur: Elément non trouvé',
  [STATUS_HTTP.SERVER_ERROR]: 'Erreur: Problème technique ! Contacter votre support',
};

/** ***********************************************************************
 * USER CONSTANTS
 ************************************************************************ */

export const ADMIN = 'Admin';
export const READER = 'Reader';
export const WRITER = 'Writer';
export const USER = 'User';
export const ALL = '';

export const PROFILS = [ADMIN, READER, WRITER, USER, ALL];

/** ***********************************************************************
 * COMMONS CONSTANTS
 ************************************************************************ */

export const MODIFIER_CLASS = {
  success: 'success',
  disabled: 'disabled',
};

/** ***********************************************************************
 * VERSION CONSTANT
 ************************************************************************ */

const VERSION_TOOLKIT = 'master';

/** ***********************************************************************
 * DEMO CONSTANTS
 ************************************************************************ */

export const STORYBOOK = 'https://axafrance.github.io/react-toolkit/latest/storybook/?path=/story/';
export const DESIGN_SYSTEM = 'https://axafrance.github.io/design-system/';
export const GITHUB = `https://github.com/AxaGuilDEv/react-toolkit/tree/v${packageJson.dependencies['@axa-fr/react-toolkit-all']}/packages/`;
export const GITHUB_API = `AxaGuilDEv/react-toolkit/${VERSION_TOOLKIT}/packages/`;
