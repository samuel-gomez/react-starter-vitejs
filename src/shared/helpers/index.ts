export { formatDate, isValidDate, setDate } from './formatDate';
export { default as isEmptyOrNull, isEmptyArray, isEmptyObject } from './isEmptyOrNull';
export { default as mergeObj } from './mergeObj';
export { default as setAnomalyEmptyItems } from './setAnomalyEmptyItems';
export { default as setConfirmClassModifier } from './setConfirmClassModifier';
export { default as emptyFunction } from './emptyFunction';
export { default as setRequiredFieldRule, setRequiredMessage } from './setRequiredFieldRule';
