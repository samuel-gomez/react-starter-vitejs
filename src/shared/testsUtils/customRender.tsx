import { render, renderHook, type RenderOptions } from '@testing-library/react';
import EnvironmentProvider from 'App/EnvironmentProvider';
import FetchProvider from 'App/FetchProvider';
import NotificationProvider from 'App/NotificationProvider';
import QueryProvider from 'App/QueryProvider';
import UserProvider from 'App/UserProvider';
import { type ReactElement, type ReactNode } from 'react';
import { MemoryRouter } from 'react-router-dom';
import MOCK_API_URL from './constants';

type TMockProvider = {
  [x: string]: Record<string, unknown | number | string> | string | boolean | null | number | Record<string, number | string>[];
};

const MockProviders =
  ({
    role = '',
    isEnabled = true,
    route = '/',
    oidcUser = { member_of: [`CN=${role}`], name: 'Samuel Gomez' },
    accessToken = 'accessToken',
    queryData,
    headers = {
      'Content-type': 'application/json; charset=UTF-8',
    },
  }: TMockProvider) =>
  ({ children }: { children: ReactNode }) => {
    const getAccessTokenFn = vi.fn().mockReturnValue(() => ({ accessToken }));
    const getUserInfosFn = vi.fn().mockReturnValueOnce(() => ({ oidcUser }));
    const useEnvFn = vi.fn().mockReturnValueOnce({
      envState: {
        environment: {
          apiUrl: MOCK_API_URL,
          fetchConfig: {
            headers,
          },
          oidc: {
            isEnabled,
          },
        },
      },
    });

    const queriesOptions = {
      retry: false,
      ...(queryData ? { queryFn: () => queryData } : {}),
    };

    return (
      <EnvironmentProvider useEnvFn={useEnvFn}>
        <UserProvider getUserInfosFn={getUserInfosFn}>
          <FetchProvider getAccessTokenFn={getAccessTokenFn}>
            <QueryProvider queriesOptions={queriesOptions}>
              <NotificationProvider>
                <MemoryRouter initialEntries={[`${route}`]}>{children}</MemoryRouter>
              </NotificationProvider>
            </QueryProvider>
          </FetchProvider>
        </UserProvider>
      </EnvironmentProvider>
    );
  };

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>, testMock: TMockProvider = { role: '' }) =>
  render(ui, { wrapper: MockProviders(testMock), ...options });

type TcustomRenderHook = (testMock?: TMockProvider) => typeof renderHook;

const customRenderHook: TcustomRenderHook = testMock => (hook, options) => renderHook(hook, { wrapper: MockProviders(testMock ?? {}), ...options });

// re-export everything
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
export { axe } from 'jest-axe';

// override render method
export { customRenderHook, customRender as render };
