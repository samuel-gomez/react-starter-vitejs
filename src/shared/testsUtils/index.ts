export * from './customRender';
export { default as emptyFunction } from '../helpers/emptyFunction';
export { default as MOCK_API_URL } from './constants';
export { default as renderWithContainer } from './renderWithContainer';
