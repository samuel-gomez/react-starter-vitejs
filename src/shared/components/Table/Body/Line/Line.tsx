import { Table as TableTk } from '@axa-fr/react-toolkit-all';
import { type ComponentPropsWithoutRef } from 'react';
import Header from '../../Headers/Header';
import { type Tcol } from '../types.d';
import Td, { type TTdContainer } from './Td';

export type TLine = ComponentPropsWithoutRef<typeof TableTk.Tr> & {
  columns: (TTdContainer & Tcol & { keyCol: string })[];
};

const Line = ({ className, columns = [], classModifier = '', children }: TLine) => (
  <TableTk.Tr classModifier={classModifier} className={className}>
    <>
      {columns.map(({ keyCol, isHeader = false, ...restTd }) =>
        isHeader ? <Header key={keyCol} scope="row" {...restTd} classModifier="row" /> : <Td key={keyCol} {...restTd} />,
      )}
      {children}
    </>
  </TableTk.Tr>
);

export default Line;
