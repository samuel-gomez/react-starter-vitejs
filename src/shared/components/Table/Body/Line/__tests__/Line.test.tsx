import { renderWithContainer } from 'shared/testsUtils';
import { within } from 'shared/testsUtils/customRender';
import Line from '../Line';

const tbodyContainer = document.createElement('tbody');

const columnsMock = [
  {
    hover: undefined,
    keyCol: 'sexe',
    label: 'F',
  },
];
const headerColumnsMock = [
  {
    ...columnsMock[0],
    isHeader: true,
  },
];

describe('Line', () => {
  it.each`
    columns              | className    | modifier      | children
    ${undefined}         | ${undefined} | ${undefined}  | ${undefined}
    ${[]}                | ${undefined} | ${undefined}  | ${undefined}
    ${columnsMock}       | ${undefined} | ${undefined}  | ${undefined}
    ${headerColumnsMock} | ${undefined} | ${undefined}  | ${undefined}
    ${columnsMock}       | ${undefined} | ${'modifier'} | ${undefined}
  `(
    'Should render <Line/> when columns: $columns, className: $className, modifier: $modifier, children: $children',
    ({ columns, className, modifier, children }) => {
      const { baseElement, container } = renderWithContainer(
        <Line className={className} columns={columns} classModifier={modifier}>
          {children}
        </Line>,
        tbodyContainer,
      );

      if (columns && columns.length > 0) {
        within(baseElement).getByText('F');

        if (columns[0]?.isHeader) {
          within(baseElement).getByRole('rowheader');
        } else {
          expect(within(baseElement).queryByRole('rowheader')).toBeNull();
        }
      } else {
        expect(within(baseElement).queryByText('F')).toBeNull();
      }

      if (modifier) {
        expect(container.firstChild).toHaveClass('af-table__tr--modifier');
      }
    },
  );
});
