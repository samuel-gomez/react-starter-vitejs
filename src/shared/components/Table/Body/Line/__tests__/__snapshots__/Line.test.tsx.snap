// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Line > Should render <Line/> when columns: [ Array(1) ], className: undefined, modifier: 'modifier', children: undefined, lineNumber: 4 1`] = `
<tbody>
  <tr
    class="af-table__tr af-table__tr--modifier"
  >
    <td
      class="af-table__cell"
    >
      F
    </td>
  </tr>
</tbody>
`;

exports[`Line > Should render <Line/> when columns: [ Array(1) ], className: undefined, modifier: undefined, children: undefined, lineNumber: 3 1`] = `
<tbody>
  <tr
    class="af-table__tr"
  >
    <td
      class="af-table__cell"
    >
      F
    </td>
  </tr>
</tbody>
`;

exports[`Line > Should render <Line/> when columns: [], className: undefined, modifier: undefined, children: undefined, lineNumber: 2 1`] = `
<tbody>
  <tr
    class="af-table__tr"
  />
</tbody>
`;

exports[`Line > Should render <Line/> when columns: undefined, className: undefined, modifier: undefined, children: undefined, lineNumber: 1 1`] = `
<tbody>
  <tr
    class="af-table__tr"
  />
</tbody>
`;
